---
title: "ProteoMill Report"
date: "`r format(Sys.time(), '%a %d %b, %Y')`"
output: html_document
---

This report was automatically generated using the [ProteoMill](https://proteomill.com/) web tool.


```{r violin1, echo = FALSE, out.width = '60%', out.height= '40%'}

if("Data summary" %in% input$p2){
	if(!is.null(maindata$udat)){
            
            dat <- make_violin()
            
            ggplot(dat, aes(x = samplename, y = values, fill = samplename)) + 
                    geom_violin(trim = FALSE, scale = "width") +
                    geom_boxplot(width=0.1, fill="white") +
                    ggthemes::theme_clean() +
                    theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust = 1), legend.position = "none") +
                    xlab("") + ylab("Log2 Expression") +
                    scale_fill_brewer(palette = "BuGn")
	}
}

```

```{r violin2, echo = FALSE, out.width = '60%', out.height= '40%'}

if("Data summary" %in% input$p2){
	if(!is.null(maindata$udat)){
            
            dat <- make_violin()
            
            ggplot(dat, aes(x = ind, y = values, fill = samplename)) + 
                geom_violin(trim = FALSE, scale = "width") +
                geom_boxplot(width=0.1, fill="white") +
                ggthemes::theme_clean() +
                theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust = 1), legend.position = "none") +
                xlab("") + ylab("Log2 Expression") +
                scale_fill_brewer(palette = "BuGn")
	}
}

```


```{r mv, echo = FALSE, out.width = '60%'}

if("Missing values" %in% input$p2) {
	if(!is.null(maindata$udat)){
		
		naf <- nafrequencies()
		
		ggplot(naf, aes(x = Var1, y = Freq, color = Var1)) +
                geom_bar(stat = "identity", width = .9, fill = "white") +
                labs(x = "Number of missing values in at least one sample", y = "Number of proteins") +
                ggthemes::theme_clean() +
                theme(axis.text.x = element_text(vjust = 0.6), legend.position = "none") +
                scale_color_grey()
	}
}

```

```{r pca2d, echo = FALSE, out.width = '60%'}

if("PCA 2D" %in% input$p2) {
	
	p.pca <- prepare_pca()
        
    si_treatment <- sampleinfo$samples$treatment
    si_replicate <- sampleinfo$samples$replicate

    mydf <- data.frame(pc1 = p.pca$x[, input$pcaDims[1]],
                       pc2 = p.pca$x[, input$pcaDims[2]],
                       si_treatment)
    
    poly.df <- mydf %>% 
            group_by(si_treatment) %>%
            do(.[chull(.$pc1, .$pc2),]) 
    
    if(input$showPolygons) {
            ggplot(mydf, aes(pc1, pc2, colour = as.factor(si_treatment))) +
                geom_polygon(data = poly.df, fill = "grey", alpha = .15) +
                geom_point(size = 5, aes(text = rownames(mydf))) +
                xlab(paste0("PC", input$pcaDims[1])) +
                ylab(paste0("PC", input$pcaDims[2])) +
                scale_color_manual(values = app_meta$palette) +
                theme_light() + theme(legend.title = element_blank()) -> p
        } else {
            ggplot(mydf, aes(pc1, pc2, colour = as.factor(si_treatment))) +
                geom_point(size = 5, aes(text = rownames(mydf))) +
                xlab(paste0("PC", input$pcaDims[1])) +
                ylab(paste0("PC", input$pcaDims[2])) +
                scale_color_manual(values = app_meta$palette) +
                theme_light() + theme(legend.title = element_blank()) -> p
        }
    

        pcaplot <- pca2d()
        ggplotly(pcaplot, tooltip = "text") %>% layout(legend = list(title=list(text='<b> Treatment </b>')))

	
	
}

```

```{r pca3d, echo = FALSE, out.width = '80%'}

if("PCA 3D" %in% input$p2) {
	
	p.pca <- prepare_pca()
	plotly::plot_ly(x = p.pca$x[,1],
	                     y = p.pca$x[,2],
	                     z = p.pca$x[,3],
	                     text = rownames(p.pca$x),
	                     hoverinfo = "text",
	                     color = sampleinfo$samples$treatment,
	                     colors = brewer.pal(length(unique(sampleinfo$samples$treatment)), "Accent")
	) %>%
	    plotly::add_markers(marker=list(size = 15, line=list(width = 1, color = "black"))) %>%
	    plotly::layout(scene = list(xaxis = list(title = paste0("PC", 1),
	                                yaxis = list(title = paste0("PC", 2),
	                                zaxis = list(title = paste0("PC", 3))))),
	                   legend = list(title=list(text='<b> Treatment </b>')))
	
}

```

```{r scree, echo = FALSE, out.width = '80%', out.height= '50%'}

if("PCA 2D" %in% input$p2 | "PCA 3D" %in% input$p2) {
	
	    p.pca <- prepare_pca()
        
        var_explained_df <- data.frame(PC = paste0("PC",1:length(sampleinfo$samples$samples)),
                                       var_explained=(p.pca$sdev)^2/sum((p.pca$sdev)^2))
        
        var_explained_df$PC <- factor(var_explained_df$PC, levels = var_explained_df$PC)
        
        var_explained_df %>%
            ggplot(aes(x = PC, y = var_explained))+
            geom_col() +
            labs(title="Scree plot: PCA on scaled data") +
            ggthemes::theme_clean() +
            theme(axis.text.x = element_text(vjust = 0.6), legend.position = "none") +
            scale_color_grey()
	
}

```

```{r heatmap, echo = FALSE, out.width = '80%'}

if("Heatmap" %in% input$p2) {
	
	dw_cor <- prepare_heatmap()
        
        if(input$showGrid) gg = 0.75
        else gg = 0
        
        heatmaply_cor(dw_cor,
                      limits = c(min(dw_cor), 1),
                      col = viridis(n = 100),
                      fontsize_row = 8,
                      fontsize_col = 8,
                      grid_gap = gg,
                      plot_method = "plotly",
                      column_text_angle = 270,
                      show_dendrogram = c(T, F),
                      row_dend_left = T
        )
        
}

```


```{r diffexpsummary, echo = FALSE}

if("Differential expression" %in% input$p2) {
	
	cbind(udat@identifiers[, ..sampleinfo$sID], udat@deoutput)
}



```

```{r diffexp_up, echo = FALSE, out.width = '80%'}

if("Differential expression" %in% input$p2) {
	
	        if(maindata$udat@deoutput[, .N] > 0) {
            diff_df <- diffv()
            
            nb.cols <- uniqueN(diff_df[, "group"])
            mycolors <- c(colorRampPalette(brewer.pal(8, "Accent"))(nb.cols))
            
            p <- plot_ly(data = diff_df,
                         type = "scatter",
                         x = ~`Fold-change`,
                         y = ~(-log10(FDR)),
                         text = ~get(sampleinfo$sID),
                         mode = "markers",
                         color = ~group,
                         colors = mycolors) %>%
                layout(yaxis = list(type = "log1p", showgrid=T,  ticks="outside", autorange=TRUE))
            p
        }
	
	
}


```


```{r upregulated_pathways, echo = FALSE, out.width = '80%'}

if("Pathway enrichment" %in% input$p2) {
	
	        if(maindata$udat@deoutput[, .N] > 0) {
            gp <- generate_pathways()
            if(gp){
                UPREGULATED_pathways <- pathways$UPREGULATED_pathways
                assign("UPREGULATED_pathways", UPREGULATED_pathways, envir = .GlobalEnv)
                DT::datatable(UPREGULATED_pathways[, -c("genes", "background")],
                              selection = 'single',
                              options = list(autoWidth = TRUE,
                                             scrollX=TRUE,
                                             columnDefs = list(
                                                 list(width = '100px', targets = c(1, 3)),
                                                 list(width = '60px', targets = c(6, 7))
                                             )))
            }

        }
	
}


```

```{r downregulated_pathways, echo = FALSE, out.width = '80%'}

if("Pathway enrichment" %in% input$p2) {
	
	if(maindata$udat@deoutput[, .N] > 0) {
            gp <- generate_pathways()
            if(gp){
            
                DOWNREGULATED_pathways <- pathways$DOWNREGULATED_pathways
                assign("DOWNREGULATED_pathways", DOWNREGULATED_pathways, envir = .GlobalEnv)
                
                DT::datatable(DOWNREGULATED_pathways[, -c("genes", "background")],
                              selection = 'single',
                              options = list(autoWidth = TRUE,
                                             scrollX=TRUE,
                                             columnDefs = list(
                                                 list(width = '100px', targets = c(1, 3)),
                                                 list(width = '60px', targets = c(6, 7))
                                             )))
            }
        }
	
}

```


